carapace-server/src/main/java/org/carapaceproxy/core/ProxyRequestsManager.java

```diff
@@ -508,25 +514,26 @@ public void reloadConfiguration(RuntimeServerConfiguration newConfiguration, Col
         @Override
         public void close() {
             connectionPools.values().forEach(connectionProvider -> {
-                connectionProvider.disposeLater().block();
+                connectionProvider.dispose(); // graceful shutdown according to disposeTimeout
             });
             connectionPools.clear();
 
             if (defaultConnectionPool != null) {
-                defaultConnectionPool.getValue()
-                        .disposeLater()
-                        .block();
-                defaultConnectionPool = null;
+                defaultConnectionPool.getValue().dispose(); // graceful shutdown according to disposeTimeout
             }
         }
 
         @Override
-        public Map.Entry<ConnectionPoolConfiguration, ConnectionProvider> apply(ProxyRequest t) {
-            String hostName = t.getRemoteAddress().getHostName();
-            return connectionPools.entrySet().stream()
+        public Map.Entry<ConnectionPoolConfiguration, ConnectionProvider> apply(ProxyRequest request) {
+            String hostName = request.getRequestHostname();
+            Map.Entry<ConnectionPoolConfiguration, ConnectionProvider> selectedPool = connectionPools.entrySet().stream()
                     .filter(e -> Pattern.matches(e.getKey().getDomain(), hostName))
                     .findFirst()
                     .orElse(defaultConnectionPool);
+
+            CarapaceLogger.debug("Using connection {0} for domain {1}", selectedPool.getKey().getId(), hostName);
```

Comment:
we should wrap this with
if (CarapaceLogger.isLoggingDebugEnabled())

Reply:
Even if not wrapped it's visible at FINE level whether debug is disabled